s/a/notify: respond with full permission set user allowed

olivercalder opened the pull request:
Temporal response caching ~~on the file descriptor~~ should be available in Ubuntu kernels ~~6.15~~ ~~6.14~~ 7.0 and later, which is currently just ~~questing and plucky~~ resolute. Test that it is working by:
- adding an integration test which opens a single file descriptor for writing, replies with "allow single", then writes to it and chmods it, and checks that the latter two operations are allowed by `apparmor` without sending a request to snapd
- adding an integration test which creates, writes, and chmods a given file path, again replying with "allow single" to the first request and checking that the latter operations are allowed by the temporal permission cache
- adding an integration test which creates, writes, and writes again to a given file path, replying with "deny single" to the first request and checking that the latter operations are denied by the temporal permission cache
- updating the existing integration tests to expect or not expect responses to be cached in the temporal cache, depending on whether it is present

~~This is expected to fail at the moment since snapd is currently conservative in the permission mask it sends back to the kernel. In particular, a reply to a request results only includes permissions which were originally requested, ANDed with the permissions corresponding to the abstract permission, if the response was "allow".~~

In order for file descriptor permission caching to have a stronger benefit, snapd needs to respond with the full set of permissions corresponding to the abstract permissions in the reply, so all those permissions will be cached.

The concern is that this could broaden the permissions beyond what the profile originally allowed. So we have to ensure that **one** of the following is true:
- AppArmor does not check the permission cache if any permissions are denied by the profile
  - This would be preferable
  - ***This is the case, all is well***
- Snapd only ever responds back with permissions which are explicitly allowed/marked as prompt in the profile
  - This would be very difficult, as different rules in the same interface may have different permissions

unskip:
- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-flag-restart
- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-integration-tests
- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-prompt-restoration
- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-smoke
- openstack:ubuntu-26.04-64:tests/main/apparmor-prompting-snapd-startup
- openstack:ubuntu-26.04-64:tests/main/interfaces-requests-activates-handlers
- openstack:ubuntu-26.04-64:tests/main/snap-interfaces-requests-control

This is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-34778 and will be tracked by https://warthogs.atlassian.net/browse/SNAPDENG-34780 once changes are added which expand the response permission masks.

olivercalder committed:
tests: add prompting integration test for FD caching with lifespan single

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

bboozzoo commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -48,6 +49,16 @@ prepare: |
+    # create_write_chmod_single requires Ubuntu 25.10 or later and go
+    if [ "$VARIANT" = create_write_chmod_single_allow ]; then
+        if os.query is-ubuntu-lt 25.10; then
+            tests.exec skip-test "Ubuntu before 25.10 does not support FD-level caching for lifespan single" && exit 0
```
were you able to confirm that the test executed on 25.10?

olivercalder replied:
Yes, the test executed on 25.10 and failed as expected. Now I'm tweaking the response permission mask from snapd to make it succeed. I'll probably retitle the PR since that will be the substantive change. Just wanted to confirm that it would fail on questing first without the change. And my local spread testing with google isn't working, and snapd garden doesn't seem to support questing yet.

olivercalder committed:
s/a/notify: respond with full permission set user allowed

Previously, a snapd allow response to a request from the kernel only
ever allowed permissions which AppArmor initially allowed and those
which were initially denied and the user then explicitly allowed.
Notably, if the user allowed a broader set of permissions, those which
were not originally requested would be omitted from the response to the
kernel.

This commit changes this behavior. Now, include all permissions the user
explicitly allowed in the allowed permission mask sent back to the
kernel.

The rationale is that newer kernels support response caching at the file
descriptor level, so if the user replies to a request prompt by allowing
with lifespan "single", the response can be cached on the FD and future
syscalls on the same file descriptor will be handled by the cache
without resulting in another request prompt to snapd.

However, permissions are exposed to users in an abstracted form, e.g.
"read", "write", "execute" for the home interface. Thus, the "write"
permission actually includes many separate operations, including
create, write, lock, chmod, etc. If an application attempts to create a
file, write to it, and then chmod it, the user will get three prompts to
"write" the same file, assuming they reply with allow "single" each
time.

By replying with the full permission mask associated with the abstract
permission which the user allowed, we can leverage the file descriptor-
level cache and handle this sequence of "write" syscalls using a single
request prompt, even if the user replies with allow "single".

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
fixup! tests: add prompting integration test for FD caching with lifespan single

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
s/a/n/listener: add debug log when listener sends response back to the kernel

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

ZeyadYasser reviewed (dismissed):
LGTM, Thanks!

olivercalder commented:
The kernel improvements to caching, including temporal cache, are targeting mid January. Once that kernel is available, we will be able to test the broader response permissions from snapd and land this PR.

In my opinion, it should be safe to land this PR already, as we know that the kernel masks out permissions which are not marked allow or prompt in the profile.

However, this is not an urgent change, I want to check with @pedronis if he has a preference about whether to land this now or wait until the kernel.

olivercalder commented:
I think the spread test needs some updating to really exercise the caching, particularly once the kernel has a temporal cache on the profile, and not just a cache on the FD. Let's hold off on landing this PR for now.

olivercalder committed:
tests: make prompting-client integration tests support temporal cache

Ubuntu kernels 7.0 and should support a temporal cache at the AppArmor
profile level. That is, if snapd sends back a reply to the kernel, that
reply will be cached for that filepath if there is a request for the
same path in the future, within the time limit.

Because of this, various prompting integration tests need to change, as
they assume e.g. that piping echo to a file will require two requests,
one for creating the file and one for writing to it, which is no longer
the case with the temporal cache.

To accomplish this, the scripts for the prompting client are now split,
with separate scripts with the `-temporal` suffix included when there
must be a difference in the expected prompts depending on whether the
temporal cache is enabled. In the test prepare, check whether the kernel
supports the temporal cache by checking for the presence of
`/proc/sys/kernel/apparmor_cache_timeout`, and if so, replaces the
default script with the script with the `-temporal` suffix.

Additionally, add several more test variants to exercise the temporal
cache.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
tests: sort the prompting-client integration test variants

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -46,16 +49,46 @@ skip:
 prepare: |
+    # Check whether kernel supports temporal cache, and whether it should, and
+    # whether the there is a temporal cache script for this variant.
+    if [ -e /proc/sys/kernel/apparmor_cache_timeout ] ; then
+        echo "Kernel supports apparmor temporal cache"
+        if [ -f "${VARIANT}-temporal.json" ] ; then
+            echo "Using script intended for kernels with temporal cache"
+            touch using-temporal-script
+            cp "${VARIANT}-temporal.json" "${VARIANT}.json"
+        else
```
There is no way a single system would wish to run with both a temporal and non-temporal script. It either supports temporal cache or it doesn't. So there's no harm in overwriting the default script with the temporal one.

olivercalder committed:
tests: fix create_write_chmod_same_path_single_allow

Without the temporal cache, there are actually five syscalls:
- One from `touch test.txt` to create the file
- Two from `echo > test.txt`, both with create|write
- Two from `chmod test.txt`, first to getattr and then chmod

With the temporal cache, there is still just one.

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder committed:
s/a/notify,tests: fix typos and improve logging

Signed-off-by: Oliver Calder <oliver.calder@canonical.com>

olivercalder commented:
If `ubuntu-daily` succeeds with the new integration test variants I believe this PR can land. From the logs I believe it's running 7.0.0-13-generic, which should have the temporal cache.

olivercalder commented:
Confirmed, it ran with kernel 7.0.0-13.

https://github.com/canonical/snapd/actions/runs/24422187976/job/71503580054?pr=15911#step:25:186

Which by the new check in prepare means it must also have the temporal cache present, else the test would have failed.

Note that the `apparmor-prompting-integration-tests` are skipped for 25.10 but the PR description `unskip`s them for 26.04. And they are confirmed to have run and passed on 26.04.

This PR is ready for review.

andrewphelpsj reviewed (approved):
Thanks!

Rnfudge02 reviewed (approved):
This looks good, thanks!

natibek commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -46,16 +49,46 @@ skip:
+        echo "Kernel supports apparmor temporal cache"
+        if [ -f "${VARIANT}-temporal.json" ] ; then
+            echo "Using script intended for kernels with temporal cache"
+            touch using-temporal-script
```
I've noticed that a lot of spread tests use files for boolean logic. Is there a reason to avoid creating an environment variable in the environment section and updating its value here?

olivercalder replied:
Uh, not a huge reason, just seems more likely a test file will be preserved even if the shell session changes for some reason. I'm not certain what of the shell environment is preserved when you transition between the `prepare` and `execute` steps, let alone doing things like `tests.session -u test exec ...`. File presence is simple and relatively bulletproof. And easier to inspect after the fact, e.g. if you shell into a spread VM with a hanging test, you can see whether the file is present, but the shell environment wouldn't be preserved.

natibek replied:
Got it. That makes sense. Though I do remember being able to print and use environment variables in the spread vm when working on the smoke tests for prompting but yeah I see how using files can be beneficial.

natibek reviewed (approved):
LGTM. Thanks!
